### ISSUE TITLE
[photospline] - divzero - loop pre-conditions aren't checked (Trac #921)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#915]
http://goo.gl/s1mdUk

loop preconditions aren't checked allowing a potential divide-by-zero error to occur in 8 steps.

fix: pre-check and hard fail if loop pre-conditions suck

Migrated from https://code.icecube.wisc.edu/ticket/921
```json
{
    "status": "closed", 
    "changetime": "2015-04-12T17:41:42", 
    "description": "http://goo.gl/s1mdUk\n\nloop preconditions aren't checked allowing a potential divide-by-zero error to occur in 8 steps.\n\nfix: pre-check and hard fail if loop pre-conditions suck", 
    "reporter": "nega", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1428860502423619", 
    "component": "combo reconstruction", 
    "summary": "[photospline] - divzero - loop pre-conditions aren't checked", 
    "priority": "normal", 
    "keywords": "photospline divzero", 
    "time": "2015-04-10T04:13:30", 
    "milestone": "", 
    "owner": "jvansanten", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20150410T11:50:02: 
* jvansanten changed resolution from "" to "fixed"
* jvansanten changed status from "new" to "closed"
* jvansanten commented:

> In changeset 131316:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="131316"
> A raft of assert()s to keep clang-analyzer happy. May fix #915 and #916.
> 
> The garbage-value warnings issued for bspline_multi.c:246 and
> convolve.c:127 appear to be garbage.
> ```

**Comment by icecube-trac** [comment]
Trac update at 20150410T21:10:06: jvansanten commented:

> In changeset 131337:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="131337"
> Suppress a bevy of clang-analyzer warnings by switching non-negative integers to
> unsigned and asserting that various unsigned counters passed to malloc() are > 0.
> Along the way, fix a few genuine memory leaks in spglam that were not obvious in
> the face its expected ~20 GB memory footprint. Also, rewrite the blossom
> convolution in a way that is slightly more memory efficient. This fixes #915 and
> #916 for reals.
> ```

**Comment by icecube-trac** [comment]
Trac update at 20150412T17:41:42: jvansanten commented:

> In changeset 131366:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="131366"
> Silence the last clang-analyzer warning, which arose because the analyzer couldn't
> tell that the loop that initialized trafo was completely equivalent to intermediate
> loop where it was used. While the warning itself was a red herring, pulling the
> calculation of the matrix elements into the flattened-inner-product loop is cleaner
> anyhow (and also done correctly this time, in marked contrast to r131337).
> 
> Fixes #915 and #916 for the 3rd or 4th time, depending on how you're counting.
> ```
